data-plane/receiver-loom/src/main/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducer.java

```diff
@@ -130,6 +130,7 @@ public Future<Void> close() {
                 }
                 sendFromQueueThread.interrupt();
                 sendFromQueueThread.join();
+                producer.flush();
```

Comment:
We call flush before closing here https://github.com/knative-extensions/eventing-kafka-broker/blob/49ad678b4a629be09ad40f9bee1a7c360837102a/data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/IngressProducerReconcilableStore.java#L253-L258

The idea of this class is to mimic the core Kafka producer interface and expose "flush" and "close" as separate methods

Reply:
Ah that makes sense - I think our issue is that we need to flush after we stop receiving requests with events, but before we close the producer. Currently, we:
1. Flush
2. Stop receiving events
3. Close the producer
This leads to a data race which causes a whole bunch of metadata requests (which aren't present now that we have this change)


I think we could fix this f we updated the producer to either:
1. Have another method on the interface to "mark the producer as closed" and then called `markedClosed -> flush -> close`, or
2. Have the `flush` method set the closed variable so that the producer stops receiving events

Reply:
> I think we could fix this f we updated the producer to either:
> 
> 1. Have another method on the interface to "mark the producer as closed" and then called `markedClosed -> flush -> close`, or
> 2. Have the `flush` method set the closed variable so that the producer stops receiving events

Isn't this exactly what the returned future communicates to the caller ?

Reply:
Yeah, that is what is happening currently. I guess what I meant was to set closed to true in the `flush` method we have in the interface, or call flush again here